src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,72 +14,766 @@
+
+    template <typename T>
+    friend constexpr bool tag_invoke(bsoncxx::detail::equal_to,
+                                     optional<T> const& opt,
+                                     nullopt_t) noexcept {
+        return !opt.has_value();
+    }
+
+    template <typename T, typename U>
+    bsoncxx_cxx14_constexpr friend auto tag_invoke(bsoncxx::detail::compare_three_way compare,
+                                                   optional<T> const& left,
+                                                   optional<U> const& right)
+        -> bsoncxx::detail::requires_t<bsoncxx::detail::strong_ordering,
+                                       bsoncxx::detail::is_totally_ordered_with<T, U>> {
+        if (left.has_value()) {
+            if (right.has_value()) {
+                return compare(*left, *right);
+            } else {
+                // non-null is greater than any null
+                return bsoncxx::detail::strong_ordering::greater;
+            }
+        } else {
+            if (right.has_value()) {
+                // Null is less than any non-null
+                return bsoncxx::detail::strong_ordering::less;
+            } else {
+                // Both are null
+                return bsoncxx::detail::strong_ordering::equal;
+            }
+        }
+    }
+
+    template <typename T, typename U>
+    bsoncxx_cxx14_constexpr friend auto tag_invoke(bsoncxx::detail::compare_three_way compare,
+                                                   optional<T> const& left,
+                                                   U const& right)
+        -> bsoncxx::detail::requires_t<bsoncxx::detail::strong_ordering,
+                                       not_an_optional<U>,
+                                       bsoncxx::detail::is_totally_ordered_with<T, U>> {
+        if (left.has_value()) {
+            return compare(*left, right);
+        }
+        // null optional is less-than any non-null value
+        return bsoncxx::detail::strong_ordering::less;
+    }
+
+    template <typename T>
+    constexpr friend bsoncxx::detail::strong_ordering tag_invoke(
+        bsoncxx::detail::compare_three_way compare, optional<T> const& left, nullopt_t) {
+        return compare(left.has_value(), false);
+    }
+};
+
+// An ADL-visible swap() should only be available for swappable objects
+template <typename T, bool IsSwappable = bsoncxx::detail::is_swappable<T>::value>
+struct optional_swap_mixin {};
+
+template <typename T>
+struct optional_swap_mixin<T, true> {
+    bsoncxx_cxx14_constexpr friend void swap(optional<T>& left, optional<T>& right) noexcept(
+        noexcept(std::is_nothrow_move_constructible<T>::value&&
+                     bsoncxx::detail::is_nothrow_swappable<T>::value)) {
+        left.swap(right);
+    }
+};
+
+// Common base class of all optionals
+template <typename T>
+class optional_common_base : optional_operators_base, optional_swap_mixin<T> {
+    using storage_type = detail::storage_for<T>;
```

Comment:
```suggestion
    using storage_type = detail::storage_for<bsoncxx::detail::remove_const_t<T>>;
```
Address problems with constructing an `optional<T const>`.